File: LeetCode/30-Day-LeetCoding-Challenge-master/Last_Stone_Weight.cpp
class Solution {
public:
    /* using vector...
     int lastStoneWeight(vector<int>& stones) {
        while(stones.size() > 1) {
            sort(stones.begin(), stones.end());
            int x = stones[stones.size()-1];
            int y = stones[stones.size()-2];
            if(x != y) {
                stones.pop_back();
                stones.pop_back();
                stones.push_back(x-y);
            } else {
                stones.pop_back();
                stones.pop_back();
            }
        }    
        if(!stones.empty()) return stones.front();
        else return 0;
    } */
    
    // using priority_queue...
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int> pq;
        for(int i=0; i<stones.size(); i++) pq.push(stones[i]);
        while(pq.size() > 1) {
            int x = pq.top(); pq.pop();
            int y = pq.top(); pq.pop();
            if(x != y) pq.push(x-y);
        }
        if(pq.size() == 1) return pq.top();
        else return 0;
    }
};

File: LeetCode/30-Day-LeetCoding-Challenge-master/Last_Stone_Weight.py
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        heap = []
        for item in stones:
            heapq.heappush(heap,(-item, item))
            
        while len(heap) > 1:
            x = heapq.heappop(heap)[1]
            y = heapq.heappop(heap)[1]
            
            if x != y:
                heapq.heappush(heap, (-(x-y), (x-y)))
        
        if len(heap) == 1:
            return heapq.heappop(heap)[1]
        else:
            return 0
